source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -94,35 +95,59 @@ void Filter::onComplete(Filters::Common::ExtAuthz::CheckStatus status) {
     break;
   case CheckStatus::Denied:
     cluster_->statsScope().counter("ext_authz.denied").inc();
-    Http::CodeUtility::ResponseStatInfo info{config_->scope(),
-                                             cluster_->statsScope(),
-                                             EMPTY_STRING,
-                                             enumToInt(Http::Code::Forbidden),
-                                             true,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             false};
+    Http::CodeUtility::ResponseStatInfo info{
+        config_->scope(),
+        cluster_->statsScope(),
+        EMPTY_STRING,
+        (update_status_code ? response->status_code : enumToInt(Http::Code::Forbidden)),
+        true,
+        EMPTY_STRING,
+        EMPTY_STRING,
+        EMPTY_STRING,
+        EMPTY_STRING,
+        false};
     Http::CodeUtility::chargeResponseStat(info);
     break;
   }
 
   // We fail open/fail close based of filter config
   // if there is an error contacting the service.
-  if (status == CheckStatus::Denied ||
-      (status == CheckStatus::Error && !config_->failureModeAllow())) {
-    Http::HeaderMapPtr response_headers{new Http::HeaderMapImpl(*getDeniedHeader())};
-    callbacks_->encodeHeaders(std::move(response_headers), true);
+  if (response->status == CheckStatus::Denied ||
+      (response->status == CheckStatus::Error && !config_->failureModeAllow())) {
+    Http::HeaderMapPtr response_headers;
+    if (update_status_code) {
```

Comment:
This logic is a bit hard to follow due to the multiple reasons we can get here combined with the above code. Comments and/or restructure would be appreciated.

Reply:
See my comment above. I will simplify this.